drv/fpga-api/src/lib.rs

```diff
@@ -306,6 +322,59 @@ pub fn load_bitstream(
     bitstream.finish_load()
 }
 
+/// Read a particular blob from auxiliary flash, hashing it along the way
+#[cfg(feature = "auxflash")]
+fn read_and_hash_blob(
+    auxflash: &mut AuxFlash,
+    blob: &AuxFlashBlob,
+    callback: &mut dyn FnMut(&[u8]) -> Result<(), FpgaError>,
+) -> Result<[u8; 32], FpgaError> {
+    let mut scratch_buf = [0u8; 128];
+    let mut pos = blob.start;
+    let mut sha = Sha3_256::new();
+    while pos < blob.end {
+        let amount = (blob.end - pos).min(scratch_buf.len() as u32);
+        let chunk = &mut scratch_buf[0..(amount as usize)];
+
+        auxflash
+            .read_slot_with_offset(blob.slot, pos, chunk)
+            .map_err(|_| FpgaError::AuxReadError)?;
+        pos += amount;
+
+        sha.update(&chunk);
+        callback(&chunk)?;
+    }
+    Ok(sha.finalize().into())
+}
+
+/// Load a bitstream from auxiliary flash
+#[cfg(feature = "auxflash")]
+pub fn load_bitstream_from_auxflash(
+    fpga: &mut Fpga,
+    auxflash: &mut AuxFlash,
+    blob: AuxFlashBlob,
+    bitstream_type: BitstreamType,
+    expected_checksum: [u8; 32],
+) -> Result<(), FpgaError> {
+    // Do an initial read to verify the checksum
```

Comment:
Reading this makes me feel we should add a `bitstream.cancel_load()`, so we can load in chunks, calculating the checksum as we go and simply canceling at the end if it doesn't match. That cancel would trigger an FPGA reset to get it ready for a retry.

I think changing the the `Bitstream` type in the API like so..
```
pub struct Bitstream(FpgaLock);

impl Bitstream {
    pub fn continue_load(&mut self, data: &[u8]) -> Result<(), FpgaError> {
        (*self.0).continue_bitstream_load(data)
    }

    pub fn finish_load(&mut self) -> Result<(), FpgaError> {
        (*self.0).finish_bitstream_load()
    }

    pub fn cancel_load(&mut self) -> Result<(), FpgaError> {
        (*self.0).reset_device()
    }
}
```
.. will do what you want. I think that gets rid of the `FpgaError::AuxTocTouChecksumMismatch` variant, and upstream code can take the mismatch error and decide what to do from there.